### PR TITLE
[DCOS-57146] Services quota table

### PR DESCRIPTION
**Pull request by brandonc**
Adds a table in the quota overview to see quota limit information for services and subgroups.

## Testing

Visit a quota tab of a quota group. You can add services and groups to the group

## Screenshots

Screen Shot 2019-08-07 at 3 40 31 PM

## Notes

- I don't know how to create a service that has tasks with mixed roles

### CODE REVIEW

File: plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx [hunk 1]
```diff
@@ -97,3 +91,47 @@ export function limitRenderer(group: ServiceGroup) {
...
+    getQuotaLimit({
+      count: servicesCount,
+      groupRoleCount
+    }),
+    servicesCount - groupRoleCount
+  ];
+}
+
+export function limitRenderer(item: ServiceGroup | ServiceTree | Service) {
+  var limit: string = "N/A";
```
Comment (GeorgiSTodorov) [inline plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx]: Can we call this something else like limitString?

File: plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx [hunk 2]
```diff
@@ -97,3 +91,47 @@ export function limitRenderer(group: ServiceGroup) {
...
+  var limitNumber: number = 0;
+
+  if (itemIsServiceGroup(item)) {
+    if (!item.quota || !item.quota.serviceRoles) {
+      return <React.Fragment />;
```
Comment (GeorgiSTodorov) [inline plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx]: Why do we have to return a React.Fragment here?

**Comment by GeorgiSTodorov** [comment]
> I don't know how to create a service that has tasks with mixed roles.

For testing, we can use fixtures I guess.

When I have a group inside a group, I can click the nested group in the table.
Peek 2019-08-08 13-24

**Comment by TattdCodeMonkey** [comment]
> When I have a group inside a group, I can click the nested group in the table and it shows the cards for the first group.

That is the behavior I would expect, the quota is on the top-level group not the sub group. We should allow navigating into sub groups so we can see the status of the services inside that group, but display the quota for the entire top-most parent group.

We can confer with @izharbuendia if that was his expectation as well.

> In services, we have one + button that has both services and groups. In quota, we have two buttons. It should be only one button.

We should fix this in a different PR, file a bug?

> After I create a service in the quota tab, it takes me back to the services tab. Is this expected?

That is also what I would expect.

File: plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx (hunk 2 above)
Comment (brandonc) [reply plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx]: If I recall, the caller interface demanded a React element

**Comment by GeorgiSTodorov** [comment]
Rebased.

**Comment by GeorgiSTodorov** [comment]
I no longer see a Quota tab when on the services overview table. It's only visible after clicking on a group.
Peek 2019-08-09 11-51

File: plugins/services/src/js/containers/services/ServiceTreeView.js [hunk 3]
```diff
@@ -123,7 +123,7 @@ class ServiceTreeView extends React.Component {
     }
     const { serviceTree } = this.props;
     const id = serviceTree.getId();
-    if (serviceTree.isRoot() || hasQuota) {
```
Comment (GeorgiSTodorov) [inline plugins/services/src/js/containers/services/ServiceTreeView.js]: Why was this removed?

File: plugins/services/src/js/structs/Service.js [hunk 4]
```diff
@@ -141,6 +141,23 @@ module.exports = class Service extends Item {
     return this.get("role") || "";
   }
 
+  getQuotaRoleStats(roleName) {
+    return (this.get("tasks") || []).reduce(
```
Comment (TattdCodeMonkey) [inline plugins/services/src/js/structs/Service.js]: Have we tested this with a `Pod`? I want to make sure `tasks` is available the same way or if we need to override there like we had to do for `getRole`

File: plugins/services/src/js/structs/Service.d.ts [hunk 5]
```diff
@@ -1,11 +1,18 @@
 import Item from "#SRC/js/structs/Item";
 import { Status } from "#PLUGINS/services/src/js/constants/ServiceStatus";
 
+interface QuotaRolesStats {
```
Comment (TattdCodeMonkey) [inline plugins/services/src/js/structs/Service.d.ts]: Should we move this to another file instead of duplicating here and in `ServiceTree.d.ts` & `ServiceGroup`?

File: plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx [hunk 6]
```diff
@@ -108,17 +108,17 @@ export function getLimitInfoForService(item: Service | ServiceTree): LimitInfo {
 
   return {
     limitText: getQuotaLimit({
-      count: stats.servicesCount,
+      count: stats.count,
```
Comment (GeorgiSTodorov) [inline plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx]: Why are we renaming this? I think that servicesCount is a lot more descriptive.

**Comment by GeorgiSTodorov** [comment]
I rebased and added the things that were missing. Feel free dropping any of my commits that are not helpful.

File: plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx (hunk 6 above)
Comment (TattdCodeMonkey) [reply plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx]: `servicesCount` isn't accurate since it could be a Pod, Service or Task that we are counting with this variable. I wondering if there is still something we can prefix `count` with to be more descriptive, or being a part of the `ServiceQuotaRolesStats` is enough?

File: plugins/services/src/js/components/ServicesQuotaOverviewTable.tsx [hunk 7]
```diff
@@ -1,83 +1,63 @@
...
-      return compatatorFor("memory");
-    case "disk":
-      return compatatorFor("disk");
-    case "gpus":
-      return compatatorFor("gpus");
+      return (a, b) => a.getName().localeCompare(b.getName());
     default:
-      return () => 0;
+      return (a, b) => {
+        const aText = getLimitInfoForService(a).limitText;
```
Comment (TattdCodeMonkey) [inline plugins/services/src/js/components/ServicesQuotaOverviewTable.tsx]: I wonder if we should be caching either the limit or the QuotaStats on the `Service` || `ServiceTree`. Counting everything every render & sort seems like a waste, but maybe that is also a pre-mature optimization.

File: plugins/services/src/js/components/ServicesQuotaOverviewTable.tsx [hunk 8]
```diff
@@ -1,83 +1,63 @@
...
   ServicesQuotaOverviewTableProps,
-  ServicesQuotaOverViewTableState
+  ServicesQuotaOverviewTableState
 > {
   constructor(props: Readonly<ServicesQuotaOverviewTableProps>) {
     super(props);
 
     this.state = {
-      groups: [],
+      items: props.serviceTree.getItems(),
```
Comment (TattdCodeMonkey) [inline plugins/services/src/js/components/ServicesQuotaOverviewTable.tsx]: Should we do a default sort on these items here?

**Comment by TattdCodeMonkey** [comment]
Integration test failures.

File: plugins/services/src/js/components/ServicesQuotaOverviewTable.tsx (hunk 7 above)
Comment (brandonc) [reply plugins/services/src/js/components/ServicesQuotaOverviewTable.tsx]: I don't think so. This is very cheap to derive and that is The React Way

File: plugins/services/src/js/components/ServicesQuotaOverviewTable.tsx (hunk 8 above)
Comment (brandonc) [reply plugins/services/src/js/components/ServicesQuotaOverviewTable.tsx]: Great catch!

File: plugins/services/src/js/structs/Service.js (hunk 4 above)
Comment (brandonc) [reply plugins/services/src/js/structs/Service.js]: Pods are supported now- insofar as they are enforced if they have a role that matches the group

**Comment by GeorgiSTodorov** [comment]
Rebased.

**Comment by GeorgiSTodorov** [comment]
I solved the issue of the invisible root table.

**Comment by TattdCodeMonkey** [comment]
This needs another rebase :/

**Comment by GeorgiSTodorov** [comment]
Rebased.

File: plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx [hunk 9]
```diff
@@ -89,11 +89,70 @@ export function limitRenderer(group: ServiceGroup) {
...
+          id={limitInfo.limitText}
           render="span"
           className="table-cell-value table-cell-flex-box"
         />
       </div>
     </TextCell>
   );
 }
+
+function itemIsServiceGroup(item: any): item is ServiceGroup {
```
Comment (pierrebeitz) [inline plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx]: that pattern can lead to runtime errors as `itemIsServiceGroup` gladly takes any input, but calls `.quota` on it, which will fail for e.g. all primitive types.

File: plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx [hunk 10]
```diff
@@ -89,11 +89,70 @@ export function limitRenderer(group: ServiceGroup) {
...
+  return item.quota && item.quota.limitStatus;
+}
+
+export function getLimitInfoForPod(item: Pod): LimitInfo {
```
Comment (pierrebeitz) [inline plugins/services/src/js/columns/QuotaOverviewLimitColumn.tsx]: is this exported intentionally? it looks like `getLimitInfoForService` is good enough as public API.

File: plugins/services/src/js/structs/Service.d.ts [hunk 11]
```diff
@@ -1,11 +1,14 @@
 import Item from "#SRC/js/structs/Item";
 import { Status } from "#PLUGINS/services/src/js/constants/ServiceStatus";
+import { ServiceQuotaRolesStats } from "../types/ServiceQuotaRolesStats";
 
 export default class Service extends Item {
   getId(): string;
   getMesosId(): string;
   getName(): string;
   getRole(): string;
+  getRootGroupName(): string;
```
Comment (pierrebeitz) [inline plugins/services/src/js/structs/Service.d.ts]: this is true, because every serviceTree or service will always contain a `/`, right?

File: plugins/services/src/js/containers/services/ServicesQuotaView.tsx [hunk 12]
```diff
@@ -82,7 +82,7 @@ class ServicesQuotaView extends React.Component<ServicesQuotaViewProps, {}> {
     }
 
     return (
-      <Page dontScroll={serviceTree.isRoot()} flushBottom={true}>
+      <Page dontScroll={false} flushBottom={true}>
```
Comment (pierrebeitz) [inline plugins/services/src/js/containers/services/ServicesQuotaView.tsx]: if `dontScroll` is always `false`, can't we delete the prop?
